# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: caps/nsJSPrincipals.cpp
# Commit: ddf4012a7652
# Full Hash: ddf4012a7652e36d144fc43bc99335276deeafbe
# Author: Christoph Kerschbaumer <ckerschb@christophkerschbaumer.com>
# Date: 2019-05-22 03:40:58
# Regressor Bug: 965637
# File Overlap Count: 1
# Description:
#   Bug 965637: Move CSP from Principal into Client, part 1: backend changes. r=mccr8
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D27654
# ==============================================================================

diff -r 6c2047bc2f3e -r ddf4012a7652 caps/nsJSPrincipals.cpp
--- a/caps/nsJSPrincipals.cpp	Tue May 21 22:51:54 2019 +0000
+++ b/caps/nsJSPrincipals.cpp	Tue May 21 23:14:27 2019 +0000
@@ -123,10 +123,10 @@
   return ReadKnownPrincipalType(aCx, aReader, tag, aOutPrincipals);
 }
 
-static bool ReadPrincipalInfo(
-    JSStructuredCloneReader* aReader, OriginAttributes& aAttrs,
-    nsACString& aSpec, nsACString& aOriginNoSuffix, nsACString& aBaseDomain,
-    nsTArray<ContentSecurityPolicy>* aPolicies = nullptr) {
+static bool ReadPrincipalInfo(JSStructuredCloneReader* aReader,
+                              OriginAttributes& aAttrs, nsACString& aSpec,
+                              nsACString& aOriginNoSuffix,
+                              nsACString& aBaseDomain) {
   uint32_t suffixLength, specLength;
   if (!JS_ReadUint32Pair(aReader, &suffixLength, &specLength)) {
     return false;
@@ -153,13 +153,14 @@
     return false;
   }
 
-  uint32_t originNoSuffixLength, policyCount;
-  if (!JS_ReadUint32Pair(aReader, &originNoSuffixLength, &policyCount)) {
+  uint32_t originNoSuffixLength, dummy;
+  if (!JS_ReadUint32Pair(aReader, &originNoSuffixLength, &dummy)) {
     return false;
   }
 
-  if (!aPolicies) {
-    MOZ_ASSERT(policyCount == 0);
+  MOZ_ASSERT(dummy == 0);
+  if (dummy != 0) {
+    return false;
   }
 
   if (!aOriginNoSuffix.SetLength(originNoSuffixLength, fallible)) {
@@ -171,29 +172,6 @@
     return false;
   }
 
-  for (uint32_t i = 0; i < policyCount; i++) {
-    uint32_t policyLength, reportAndMeta;
-    if (!JS_ReadUint32Pair(aReader, &policyLength, &reportAndMeta)) {
-      return false;
-    }
-    bool reportOnly = reportAndMeta & 1;
-    bool deliveredViaMetaTag = reportAndMeta & 2;
-
-    nsAutoCString policyStr;
-    if (!policyStr.SetLength(policyLength, fallible)) {
-      return false;
-    }
-
-    if (!JS_ReadBytes(aReader, policyStr.BeginWriting(), policyLength)) {
-      return false;
-    }
-
-    if (aPolicies) {
-      aPolicies->AppendElement(ContentSecurityPolicy(
-          NS_ConvertUTF8toUTF16(policyStr), reportOnly, deliveredViaMetaTag));
-    }
-  }
-
   uint32_t baseDomainIsVoid, baseDomainLength;
   if (!JS_ReadUint32Pair(aReader, &baseDomainIsVoid, &baseDomainLength)) {
     return false;
@@ -259,9 +237,7 @@
     nsAutoCString spec;
     nsAutoCString originNoSuffix;
     nsAutoCString baseDomain;
-    nsTArray<ContentSecurityPolicy> policies;
-    if (!ReadPrincipalInfo(aReader, attrs, spec, originNoSuffix, baseDomain,
-                           &policies)) {
+    if (!ReadPrincipalInfo(aReader, attrs, spec, originNoSuffix, baseDomain)) {
       return false;
     }
 
@@ -275,7 +251,7 @@
 
     // XXX: Do we care about mDomain for structured clone?
     aInfo = ContentPrincipalInfo(attrs, originNoSuffix, spec, Nothing(),
-                                 std::move(policies), baseDomain);
+                                 baseDomain);
   } else {
 #ifdef FUZZING
     return false;
@@ -344,37 +320,23 @@
   return true;
 }
 
-static bool WritePrincipalInfo(
-    JSStructuredCloneWriter* aWriter, const OriginAttributes& aAttrs,
-    const nsCString& aSpec, const nsCString& aOriginNoSuffix,
-    const nsCString& aBaseDomain,
-    const nsTArray<ContentSecurityPolicy>* aPolicies = nullptr) {
+static bool WritePrincipalInfo(JSStructuredCloneWriter* aWriter,
+                               const OriginAttributes& aAttrs,
+                               const nsCString& aSpec,
+                               const nsCString& aOriginNoSuffix,
+                               const nsCString& aBaseDomain) {
   nsAutoCString suffix;
   aAttrs.CreateSuffix(suffix);
-  size_t policyCount = aPolicies ? aPolicies->Length() : 0;
 
   if (!(JS_WriteUint32Pair(aWriter, suffix.Length(), aSpec.Length()) &&
         JS_WriteBytes(aWriter, suffix.get(), suffix.Length()) &&
         JS_WriteBytes(aWriter, aSpec.get(), aSpec.Length()) &&
-        JS_WriteUint32Pair(aWriter, aOriginNoSuffix.Length(), policyCount) &&
+        JS_WriteUint32Pair(aWriter, aOriginNoSuffix.Length(), 0) &&
         JS_WriteBytes(aWriter, aOriginNoSuffix.get(),
                       aOriginNoSuffix.Length()))) {
     return false;
   }
 
-  for (uint32_t i = 0; i < policyCount; i++) {
-    nsCString policy;
-    CopyUTF16toUTF8((*aPolicies)[i].policy(), policy);
-    uint32_t reportAndMeta =
-        ((*aPolicies)[i].reportOnlyFlag() ? 1 : 0) |
-        ((*aPolicies)[i].deliveredViaMetaTagFlag() ? 2 : 0);
-    if (!(JS_WriteUint32Pair(aWriter, policy.Length(), reportAndMeta) &&
-          JS_WriteBytes(aWriter, PromiseFlatCString(policy).get(),
-                        policy.Length()))) {
-      return false;
-    }
-  }
-
   if (aBaseDomain.IsVoid()) {
     return JS_WriteUint32Pair(aWriter, 1, 0);
   }
@@ -413,8 +375,7 @@
   const ContentPrincipalInfo& cInfo = aInfo;
   return JS_WriteUint32Pair(aWriter, SCTAG_DOM_CONTENT_PRINCIPAL, 0) &&
          WritePrincipalInfo(aWriter, cInfo.attrs(), cInfo.spec(),
-                            cInfo.originNoSuffix(), cInfo.baseDomain(),
-                            &(cInfo.securityPolicies()));
+                            cInfo.originNoSuffix(), cInfo.baseDomain());
 }
 
 bool nsJSPrincipals::write(JSContext* aCx, JSStructuredCloneWriter* aWriter) {